### PR TITLE
make: Add ineffassign as a part of precheck

### LINKED ISSUE #3609: Include ineffassign in tests
To avoid issues like #3577 in future, we can try to include `ineffassign` in our tests. I.e. as a part of `make tests` target.

In order to do that, we need to include https://github.com/gordonklaus/ineffassign in https://github.com/cilium/packer-ci-build.

### PR DESCRIPTION
**Pull request by vadorovsky**
Fixes #3609

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4374)

### CODE REVIEW

**Comment by vadorovsky** [comment]
test-me-please

File: pkg/identity/identity.go [hunk 1]
```diff
@@ -127,7 +127,7 @@ func (pair *IPIdentityPair) IsHost() bool {
 func (pair *IPIdentityPair) PrefixString() string {
 	var suffix string
 	if !pair.IsHost() {
-		ones := net.IPv6len
+		var ones int
```
Comment (joestringer) [inline pkg/identity/identity.go]: This case looks interesting to me actually, there's two broad cases: Mask is nil, or not... then two inner cases: IPv4 or IPv6. It looks like if it's not IPv4, we could end up dereferencing a nil Mask.

Perhaps it should be something more like:

```
var ones int
if pair.Mask == nil {
    if pair.IP.To4() != nil {
        ones = net.IPv4len
    } else {
        ones = net.IPv6len
    }
} else {
    ones, _ = pair.Mask.Size()
}
```

File: pkg/kvstore/etcd.go [hunk 2]
```diff
@@ -580,7 +580,7 @@ reList:
 					}
 
 					// Retrieve latest revision
-					lastRev = 0
+					// lastRev = 0
```
Comment (joestringer) [inline pkg/kvstore/etcd.go]: It looks to me like this should be reused at the start of the `reList` loop, so it's not really ineffassign. Maybe we can find a way to teach ineffassign about this?

File: pkg/workloads/containerd.go [hunk 3]
```diff
@@ -106,7 +106,7 @@ func newContainerDClient(opts workloadRuntimeOpts) (WorkloadRuntime, error) {
 		ep = "unix://" + ep
 	}
 	rsc, err := newCRIClient(context.WithValue(context.Background(), epOpt, ep))
-	return &containerDClient{c, rsc}, nil
```
Comment (joestringer) [inline pkg/workloads/containerd.go]: Nice catch.

File: test/runtime/Policies.go [hunk 4]
```diff
@@ -1562,7 +1562,7 @@ var _ = Describe("RuntimeValidatedPolicyImportTests", func() {
 		Expect(httpd1EndpointModel.Status.Identity).To(Not(BeNil()), "Expected non-nil identity for endpoint %s", helpers.Httpd1)
 		Expect(httpd1EndpointModel.Status.Policy).To(Not(BeNil()), "Expected non-nil policy for endpoint %s", helpers.Httpd1)
 
-		httpd1SecurityIdentity = httpd1EndpointModel.Status.Identity.ID
+		// httpd1SecurityIdentity = httpd1EndpointModel.Status.Identity.ID
```
Comment (joestringer) [inline test/runtime/Policies.go]: Why is this redundant but the following line isn't?

**Review by joestringer — COMMENTED** [review]
Looks like one of the cases picked up a nice bug, another one of the cases I'm not really sure that the ineffassign tool got it right. I didn't look closely at the CRI changes.

**Review by aanm — CHANGES_REQUESTED** [review]
Please split the CRI related changes to be in a new PR so we can push it to 1.1

**Comment by joestringer** [comment]
^ Same for the `PrefixString()` bug, that should go into 1.1 as well.

File: pkg/identity/identity.go (hunk 1 above)
Comment (vadorovsky) [reply pkg/identity/identity.go]: Yes, that looks reasonable. I also had a feeling that the if statement in the current shape is a little bit unclear.

File: pkg/kvstore/etcd.go (hunk 2 above)
Comment (vadorovsky) [reply pkg/kvstore/etcd.go]: I think it will be hard to teach ineffassign about this (it just doesn't have a support of skipping errors). But I have an another idea - we can wrap `ineffassign` call in `contrib/scripts/check-ineffassign.sh` and do there something like:

```
ineffassign $TOPDIR | grep -v "# ineffassign: ignore"
````

Because ineffassign output shows the whole line.

File: test/runtime/Policies.go (hunk 4 above)
Comment (vadorovsky) [reply test/runtime/Policies.go]: Because `httpd2SecurityIdentity` variable is used in line 1569. `httpd1...` variable isn't used by the rest of the function at all.

**Review by vadorovsky — COMMENTED** [review]
@aanm @joestringer Separate pull requests created:
https://github.com/cilium/cilium/pull/4385
https://github.com/cilium/cilium/pull/4386

**Comment by vadorovsky** [comment]
test-me-please

**Comment by vadorovsky** [comment]
@aanm @joestringer PTAL

I added a wrapper script which ignores an error if a line contains `// ineffassign: ignore` comment.

**Review by aanm — APPROVED** [review]

File: pkg/bpf/bpffs.go [hunk 5]
```diff
@@ -258,7 +258,7 @@ func checkOrMountDefaultLocations() error {
 			return err
 		}
 		if !cMounted {
-			err = mountFS()
+			err = mountFS() // ineffassign: ignore
```
Comment (joestringer) [inline pkg/bpf/bpffs.go]: Do you think that this ineffassign is because of the `... err := isBPFFS()` on line 256?

Maybe that line should be updated to not declare `err`. In that case, this may also be a bug and so the fix should be split into a separate PR and we can make sure it goes into 1.1.

**Review by joestringer — COMMENTED** [review]
Looks good, one comment to follow up below. Also, could you update the developer documentation to describe how to install ineffassign since it'll be a build-time dependency?

**Comment by joestringer** [comment]
Looks like the build failure is relevant, the new dependency also needs to be added to the vagrant files for the tests:

```
08:43:53     runtime: contrib/scripts/check-ineffassign.sh: line 6: ineffassign: command not found
08:43:53     runtime: Makefile:212: recipe for target 'ineffassign' failed
08:43:53     runtime: make: *** [ineffassign] Error 1
08:43:53 The SSH command responded with a non-zero exit status. Vagrant
08:43:53 assumes that this means the command failed. The output for this command
08:43:53 should be in the log above. Please read the output to determine what
08:43:53 went wrong.
08:43:53 
08:43:53         ===================== ERROR - VM PROVISION FAILED =====================
08:43:53 
08:43:53         Unable to provision and start VM "runtime": exit status 1", vm, err
08:43:53 
08:43:53         =======================================================================
```

File: pkg/bpf/bpffs.go (hunk 5 above)
Comment (raybejjani) [reply pkg/bpf/bpffs.go]: That might be it. Trying this out https://play.golang.org/p/S8liQ6Y4UsM without the if produced output as expected. I had thought assignments like this didn't shadow already existing variables unless it was in the `if` setup section (`if z=1; z < 9`).

I think I would prefer to refactor this function a little. Either use a big switch where each case corresponds to the various combinations of `mounted` and `bpffsInstance` or return early when possible. So, an early return on line 231, another early return when `mounted && bpffsInstance`, the implied default here. I'm not sure what the precedence is between `!mounted` and `!bpffsInstance` is  but both of those could return internally directly, since they are simple enough.

File: pkg/kvstore/etcd.go [hunk 6]
```diff
@@ -580,7 +580,7 @@ reList:
 					}
 
 					// Retrieve latest revision
-					lastRev = 0
+					lastRev = 0 // ineffassign: ignore
```
Comment (raybejjani) [inline pkg/kvstore/etcd.go]: Could we just move this line and/or the declaration of lastRev to the beginning of the `reList` loop? That's the only way we would end up with lastRev=0 anyway... I might have missed something, this code is incredibly difficult to follow.
While supporting ignore is handy, it defeats the point of using ineffassign. The tool is telling us that we've done something unintended in the language so our code is very likely incorrect.

File: pkg/bpf/bpffs.go (hunk 5 above)
Comment (vadorovsky) [reply pkg/bpf/bpffs.go]: Yes, you are right. I also didn't expect that `v, err := something()` shadows the previously declared `err` variable, but it seems tat it does. I will early return in the separate PR.

File: pkg/kvstore/etcd.go (hunk 6 above)
Comment (vadorovsky) [reply pkg/kvstore/etcd.go]: You're right, I will try that and submit a separate PR if moving that assignment to the beginning of `reList` will work

Comment (raybejjani) [reply pkg/kvstore/etcd.go]: Thanks. To be clear, I'm not opposed to this PR being merged, but in the long run I'm worried the ignore will allow some bad behaviour on our part ("Oh, I just need to merge this to make this deadline" and then forgetting about it).

**Comment by vadorovsky** [comment]
@joestringer @raybejjani 
#4501 
#4502

**Comment by vadorovsky** [comment]
I did a rebase, but we also need ineffassign in Vagrant box https://github.com/cilium/packer-ci-build/pull/67

**Comment by vadorovsky** [comment]
test-me-please

**Comment by joestringer** [comment]
@mrostecki Now that ineffassign is in the vagrant image, can this be rebased?

**Comment by vadorovsky** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

File: pkg/kvstore/etcd.go [hunk 7]
```diff
@@ -298,7 +298,10 @@ func newEtcdClient(config *client.Config, cfgPath string) (BackendOperations, er
 		// Run renewLease once the firstSession is received
 		if err := ec.renewLease(); err != nil {
 			c.Close()
-			err = fmt.Errorf("unable to create default lease: %s", err)
+			if err = fmt.Errorf("unable to create default lease: %s", err); err != nil {
+				log.Fatal(err)
```
Comment (joestringer) [inline pkg/kvstore/etcd.go]: These lines could be de-duplicated by changing the if statement instead to:

```
err = ec.renewLease()
if err != nil {
...
```

This way the existing `log.Fatal()` would be hit, like the previous code intended.

File: test/k8sT/Policies.go [hunk 8]
```diff
@@ -661,6 +661,7 @@ var _ = Describe("K8sPolicyTest", func() {
 			Expect(err).To(BeNil(), "Cannot get web pods")
 
 			serviceIP, port, err := kubectl.GetServiceHostPort(helpers.DefaultNamespace, "redis-master")
+			Expect(err).To(BeNil())
```
Comment (joestringer) [inline test/k8sT/Policies.go]: Please add a string comment to describe the problem here, similar to how line 661 has.

File: test/runtime/chaos.go [hunk 9]
```diff
@@ -114,6 +114,7 @@ var _ = Describe("RuntimeChaos", func() {
 		status.ExpectFail("leftover interface were not properly cleaned up")
 
 		links, err := vm.Exec("sudo ip link show | wc -l").IntOutput()
+		Expect(err).Should(BeNil())
```
Comment (joestringer) [inline test/runtime/chaos.go]: Again, please add an error string:

```
Expect(err).To(BeNil(), "Unable to list IPs")
```

File: test/runtime/connectivity.go [hunk 10]
```diff
@@ -263,6 +263,7 @@ var _ = Describe("RuntimeConnectivityTest", func() {
 			err = helpers.RenderTemplateToFile(policyFileName, policy, os.ModePerm)
 			Expect(err).Should(BeNil())
 			_, err = vm.PolicyImportAndWait(helpers.GetFilePath(policyFileName), helpers.HelperTimeout)
+			Expect(err).Should(BeNil())
```
Comment (joestringer) [inline test/runtime/connectivity.go]: Please add error string.

**Review by joestringer — COMMENTED** [review]
Some of the ineffassigns that are showing up in the testsuite show some really useful cases where we're unintentionally neglecting the error conditions. Looking forward to getting this in!

Please just add some error strings to the new expectations, it's difficult to debug what goes wrong when the test fails and doesn't have any string associated with the error.

**Comment by joestringer** [comment]
Weird, we don't seem to have ineffassign in the CI?

```
09:51:40 contrib/scripts/check-ineffassign.sh
09:51:40 ++ git rev-parse --show-toplevel
09:51:40 + TOPDIR=/go/src/github.com/cilium/cilium
09:51:40 + ineffassign /go/src/github.com/cilium/cilium
09:51:40 contrib/scripts/check-ineffassign.sh: line 7: ineffassign: command not found
```

**Comment by vadorovsky** [comment]
@joestringer That's indeed weird. Do we have any CI tests running outside Vagrant? Vagrant box for sure has `ineffassign`.

Dev VM:

```
 mrostecki   ineffassign  …  github.com  cilium  cilium  vagrant ssh
[...]
vagrant@runtime1:~/go/src/github.com/cilium/cilium$ which ineffassign
/home/vagrant/go/bin/ineffassign
```

Ginkgo VM:

```
 mrostecki   ineffassign  …  cilium  cilium  test  vagrant ssh runtime
[...]
vagrant@runtime:~$ which ineffassign
/home/vagrant/go/bin/ineffassign
```

**Comment by joestringer** [comment]
@mrostecki Good point, it's in the precheck stage:
https://jenkins.cilium.io/view/PR/job/Cilium-PR-Ginkgo-Tests-Validated/5732/flowGraphTable/

Maybe we need to put this somewhere else in the Jenkins scripts or something. @eloycoto / @ianvernon ?

**Review by joestringer — COMMENTED** [review]
Just need to figure out why the CI "precheck" stage doesn't have ineffassign available and make it available.

**Comment by vadorovsky** [comment]
We need to wait for https://github.com/cilium/cilium/pull/5245 and new cilium-builder image

**Review by joestringer — APPROVED** [review]
Just need to rebase against master when the builder-image has been updated to include `ineffassign`.

**Comment by vadorovsky** [comment]
test-me-please

**Comment by vadorovsky** [comment]
test-me-please

**Comment by vadorovsky** [comment]
Seems unrelated

```
11:16:11 [K8s-1.8] • Failure in Spec Setup (BeforeEach) [434.414 seconds]
11:16:11 [K8s-1.8] K8sKafkaPolicyTest
11:16:11 [K8s-1.8] /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
11:16:11 [K8s-1.8]   Kafka Policy Tests
11:16:11 [K8s-1.8]   /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:344
11:16:11 [K8s-1.8]     KafkaPolicies [BeforeEach]
11:16:11 [K8s-1.8]     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:376
11:16:11 [K8s-1.8] 
11:16:11 [K8s-1.8]     Kafka Pods are not ready after timeout
11:16:11 [K8s-1.8]     Expected
11:16:11 [K8s-1.8]         <*errors.errorString | 0xc420676120>: {
11:16:11 [K8s-1.8]             s: "Timeout reached: timed out waiting for pods with filter -l zgroup=kafkaTestApp to be ready",
11:16:11 [K8s-1.8]         }
11:16:11 [K8s-1.8]     to be nil
11:16:11 [K8s-1.8] 
11:16:11 [K8s-1.8]     /home/jenkins/workspace/Cilium-PR-Ginkgo-Tests-Validated/src/github.com/cilium/cilium/test/k8sT/KafkaPolicies.go:133
```

**Comment by vadorovsky** [comment]
test-me-please
